File: lintcode-training/SubsetsII.java
class Solution {
    /**
     * @param nums: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public ArrayList<ArrayList<Integer>> subsetsWithDup(int[] nums) {
        // write your code here
        ArrayList<ArrayList<Integer>> results = new ArrayList<>();
        if (nums == null) return results;
        
        if (nums.length == 0) {
            results.add(new ArrayList<Integer>());
            return results;
        }
        Arrays.sort(nums);

        ArrayList<Integer> subset = new ArrayList<>();
        helper(nums, 0, subset, results);
        
         return results;
        
        
    }
    public void helper(int[] nums, int startIndex, ArrayList<Integer> subset, ArrayList<ArrayList<Integer>> results){
        results.add(new ArrayList<Integer>(subset));
        for(int i=startIndex; i<nums.length; i++){
            if (i != startIndex && nums[i]==nums[i-1]) {
                continue;
            }
            subset.add(nums[i]);
            helper(nums, i+1, subset, results);
            subset.remove(subset.size()-1);
        }
    }
}


// class SubsetsII extends Solution{
//     public static void main(int[] nums){
//         SubsetsII a1=new SubsetsII();
//         a1.subsetsWithDup([1, 2, 2]);

//     }

// }

File: lintcode-training/SubsetsII.cpp
#include<vector>
#include<algorithm>
using namespace std;


class Solution {
public:
    vector<vector<int> > result;
    bool equls( vector<int> a, vector<int> b){
        if (a.size() != b.size())
            return false;
        int cnt = a.size();
        for (int i = 0; i < cnt;++i)
            if (a[i]!=b[i]) return false;
        return true;
    }
    void dfs(vector<int> tmp, int x , vector<int> nums) {
        if (x == nums.size()) {
            int cnt = result.size();
            for ( int i = 0; i  < cnt ; ++i)
                if ( equls(result[i], tmp) )
                        return ;
            result.push_back(tmp);
            return ;
        }
        dfs(tmp, x + 1, nums);
        tmp.push_back(nums[x]);
        dfs(tmp, x + 1, nums);
    }
    vector<vector<int> > subsetsWithDup(vector<int> &nums){
        sort(nums.begin(), nums.end());
    	vector<int> tmp;
    	dfs(tmp, 0 , nums) ;
    	// write your code here
    	return result;
    }
};

//int main(){
//    Solution result;
//
//    result.subsetsWithDup([1, 2, 2]);
//}